src/main/java/pl/javastart/streamstask/StreamsTask.java

```diff
@@ -38,22 +39,35 @@ private void run() {
 
     // metoda powinna zwracać listę kobiet (sprawdzając, czy imię kończy się na "a")
     Collection<User> findWomen(Collection<User> users) {
-        throw new RuntimeException("Not implemented");
+        return users.stream()
+                .filter(user -> user.getName().endsWith("a"))
+                .collect(Collectors.toList());
     }
 
     // metoda powinna zwracać średni wiek mężczyzn (sprawdzając, czy imię nie kończy się na "a")
     Double averageMenAge(Collection<User> users) {
-        throw new RuntimeException("Not implemented");
+        return users.stream()
+                .filter(user -> !user.getName().endsWith("a"))
+                .collect(Collectors.averagingDouble(User::getAge));
     }
 
     // metoda powinna zwracać wydatki zgrupowane po ID użytkownika
     Map<Long, List<Expense>> groupExpensesByUserId(Collection<User> users, List<Expense> expenses) {
-        throw new RuntimeException("Not implemented");
+        return expenses.stream()
+                .collect(Collectors.groupingBy(Expense::getUserId));
     }
 
     // metoda powinna zwracać wydatki zgrupowane po użytkowniku
     // podobne do poprzedniego, ale trochę trudniejsze
     Map<User, List<Expense>> groupExpensesByUser(Collection<User> users, List<Expense> expenses) {
-        throw new RuntimeException("Not implemented");
+        return expenses.stream()
+                .collect(Collectors.groupingBy(expense -> findUserById(users, expense.getUserId())));
+    }
+
+    private User findUserById(Collection<User> users, Long userId) {
+        return users.stream()
+                .filter(user -> user.getId().equals(userId))
+                .findFirst()
+                .orElse(null);
     }
```

Comment:
Tutaj zrobiła się złożoność obliczeniowa O(n^2), bo dla każdego wydatku przeszukujesz listę użytkowników. Lepiej rozbić to na dwa etapy: najpierw przerobić listę użytkowników na mapę i z niej korzystać do wyszukiwnia.